File: rental_store/frontend/src/services/invitationService.ts
import { http } from "./apiClient";
import type { Invitation, InvitationKind } from "../types/invitation";

interface InvitationResponse {
  id: string;
  token: string;
  kind: InvitationKind;
  email: string | null;
  status: Invitation["status"];
  createdAt: string;
  expiresAt: string;
}

interface RedeemInvitationInput {
  name: string;
  email?: string;
  password: string;
  businessName?: string;
}

function toInvitation(raw: InvitationResponse): Invitation {
  return {
    id: raw.id,
    kind: raw.kind,
    token: raw.token,
    ...(raw.email ? { email: raw.email } : {}),
    status: raw.status,
    createdAt: raw.createdAt,
    expiresAt: raw.expiresAt,
  };
}

export const invitationService = {
  async getByToken(token: string): Promise<Invitation> {
    const raw = await http.get<InvitationResponse>(
      `/invitations/token/${encodeURIComponent(token)}`,
    );
    return toInvitation(raw);
  },

  async redeem(token: string, input: RedeemInvitationInput): Promise<void> {
    await http.post<void>(
      `/invitations/token/${encodeURIComponent(token)}/redeem`,
      input,
    );
  },
};

File: rental_store/backend/src/main/java/com/pata/keja/service/InvitationService.java
package com.pata.keja.service;

import com.pata.keja.enums.InvitationKind;
import com.pata.keja.enums.InvitationStatus;
import com.pata.keja.dto.invitation.InvitationCreateRequest;
import com.pata.keja.dto.invitation.InvitationResponse;
import com.pata.keja.dto.invitation.InvitationSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface InvitationService {

    Page<InvitationSummaryResponse> list(InvitationStatus status,
            InvitationKind kind,
            Pageable pageable);

    InvitationResponse getById(String invitationId);

    /** Look up by token — used by the public "redeem invitation" page. */
    InvitationResponse getByToken(String token);

    InvitationResponse create(String adminId, InvitationCreateRequest req);

    /** Return the active caretaker link for this hostel, or create one. */
    InvitationResponse createForLandlord(String landlordId, String hostelId);

    InvitationResponse revoke(String invitationId, String adminId);

    /**
     * Redeem an invitation. Called after the invitee submits their details
     * (name, password). Creates the appropriate user, marks the invitation
     * as USED, and returns the new user's id (as a plain string so we don't
     * leak entity classes through the service interface).
     */
    String redeem(String token, String name, String email, String password, String businessName);

    int expireOverdueInvitations();
}
